### PR TITLE
Implement Sublattice KKS phase field model

### LINKED ISSUE #5862: Support for CALPHAD free energies with sublattice compositions
In #5787 a small code based on pycalphad was added to MOOSE. This code allows exporting free energy expressions from TDB files for use in MOOSE. CALPHAD uses the concept of _sublattice compositions_, whereas the phase field equations operate on the _total composition_ fields. Two problems need to be solved:
- [ ] Constrained minimization of the free energy density with respect to the sublattice compositions
- [ ] Calculate chemical potentials (i.e. derivatives of the free energy w.r.t. the total concentrations)

### PR DESCRIPTION
**Pull request by dschwen**
This implements the sublattice kim-kim-suzuki (SLKKS) phase field model from

Schwen, D., C. Jiang, and L. K. Aagesen. "A sublattice phase-field model for direct CALPHAD database coupling." Computational Materials Science 195 (2021): 110466.

Permission was granted by the author to distribute the included `CrFe_Jacob.tdb` thermodynamic database file.

Closes #5862

### CODE REVIEW

**Comment by dschwen** [comment]
Here is the notebook with the current status (phase diagram and plots of free energy and sublattice concentrations)

https://github.com/idaholab/moose/blob/070172ca7ea78977d872bf2ac2b6d2102bfbb60a/modules/phase_field/examples/slkks/CrFe.ipynb

File: modules/phase_field/test/tests/KKS_system/auxkernel.i [hunk 1]
```diff
@@ -0,0 +1,263 @@
+#
+# This test checks if the thwo phase and lagrange multiplier solutions can be replicated
```
Comment (laagesen) [inline modules/phase_field/test/tests/KKS_system/auxkernel.i:2]: typo thwo->two

File: modules/phase_field/test/tests/KKS_system/lagrange_multiplier.i [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+[]
+
+[Preconditioning]
+  active = 'full'
+  [full]
+    type = SMP
+    full = true
+  []
+  [mydebug]
+    type = FDP
```
Comment (laagesen) [inline modules/phase_field/test/tests/KKS_system/lagrange_multiplier.i:323]: Any reason to include this? Looks like `active = 'full'` is setting the previous block to be active, so can we remove the `mydebug` block?

File: modules/phase_field/test/tests/slkks/tests [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    design = 'SLKKS.md'
+    requirement = 'The system shall be able to solve for mass transport, phases, and sublattice concentrations simultaneously in the SLKKS model.'
+  []
+  [weighted_average]
+    type = Exodiff
+    input = weighted_average.i
+    exodiff = weighted_average_out.e
+    issues = '#5862'
+    design = 'WeightedVariableAverage.md'
+    requirement = 'The system shall be able to compute teh average value of a variable field weighted by a material property.'
```
Comment (laagesen) [inline modules/phase_field/test/tests/slkks/tests:24]: typo teh->the

**Review by laagesen — CHANGES_REQUESTED** [review]

File: modules/phase_field/src/kernels/SLKKSMultiACBulkC.C [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+#include "SLKKSMultiACBulkC.h"
+
+registerMooseObject("PhaseFieldApp", SLKKSMultiACBulkC);
+
+InputParameters
+SLKKSMultiACBulkC::validParams()
+{
+  auto params = SLKKSMultiPhaseBase::validParams();
+  params.addClassDescription("Multi-phase KKS model kernel (part 2 of 2) for the bulk Allen-Cahn. "
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/phase_field/src/kernels/SLKKSMultiACBulkC.C:18]: Yeah, this is misleading, and probably a result of copy paste. The other kernel is just `KKSMultiACBulkF`.

File: modules/phase_field/include/kernels/SLKKSMultiPhaseBase.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+#include "KernelValue.h"
+#include "JvarMapInterface.h"
+#include "DerivativeMaterialInterface.h"
+
+/**
+ * Enforce sum of phase sublattice concentrations to be the real concentration.
+ * D. Schwen et al. https://doi.org/10.1016/j.commatsci.2021.110466
+ *
+ * \see SLKKSPhaseChemicalPotential
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/phase_field/include/kernels/SLKKSMultiPhaseBase.h:20]: https://www.doxygen.nl/manual/commands.html#cmdsee

File: modules/phase_field/test/tests/KKS_system/lagrange_multiplier.i (hunk 2 above)
Comment (dschwen) [reply modules/phase_field/test/tests/KKS_system/lagrange_multiplier.i:323]: that entire `[Preconditioning]` block is unneeded. NEWTON creates a full matrix by default now.

File: modules/phase_field/examples/slkks/CrFe.i [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+    function_name = g2
+    g_order = SIMPLE
+    eta = eta2
+  []
+
+  # constant properties
+  [constants]
+    type = GenericConstantMaterial
+    prop_names = 'D   L   kappa'
+    prop_values = '10  1   0.1  '
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/phase_field/examples/slkks/CrFe.i:294]: Hm, converges _a lot_ worse that way. Odd.

**Review by laagesen — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]

**Pull request merged**